### PR TITLE
Refactor `useHistoryVersions` and `useRoomNotificationSettings`

### PR DESCRIPTION
**Pull request by nimeshnayaju**

### CODE REVIEW

File: packages/liveblocks-react/src/umbrella-store.ts [hunk 1]
```diff
@@ -299,6 +299,72 @@ const noop = Promise.resolve();
 
 const ASYNC_LOADING = Object.freeze({ isLoading: true });
 
+class SingleResource {
+  public readonly observable: Observable<void>;
+  private _eventSource: EventSource<void>;
+  private _fetchPage: () => Promise<void>;
+  private _cachedPromise: UsablePromise<void> | null = null;
```
Comment (nimeshnayaju) [inline packages/liveblocks-react/src/umbrella-store.ts:306]: I've duplicated the implementation from `PaginatedResource` for now but we most likely need an abstraction that both single and paginated resources can extends.

Comment (nvie) [reply packages/liveblocks-react/src/umbrella-store.ts:306]: Yep! Duplication isn't bad here. Premature DRYing up or using subclasses would be much worse. Until the abstraction becomes apparent, we can keep the duplication. If they're still somewhat symmetric, this is fine.

**Review by nvie — APPROVED** [review]
Looking good! 👌

File: packages/liveblocks-react/src/umbrella-store.ts [hunk 2]
```diff
@@ -299,6 +299,72 @@ const noop = Promise.resolve();
 
 const ASYNC_LOADING = Object.freeze({ isLoading: true });
 
+class SingleResource {
```
Comment (nvie) [inline packages/liveblocks-react/src/umbrella-store.ts:302]: Maybe `SinglePageResource`?

**Comment by nvie** [comment]
@nimeshnayaju I just back-merged `pagination-cleanups` into this PR, and resolved [one conflict](https://github.com/liveblocks/liveblocks/pull/1989/commits/10c7a3692319f3da1961eccba52ddcc6bf0019b4) caused by #1995 🙏

File: packages/liveblocks-core/src/room.ts [hunk 3]
```diff
@@ -1646,7 +1650,32 @@ export function createRoom<
...
+        }
+      );
+
+      return {
+        versions,
+      };
+    } else {
+      throw new Error(
+        `There was an error while getting text versions for room ${config.roomId}`
+      );
```
Comment (nvie) [inline packages/liveblocks-core/src/room.ts:1677]: @nimeshnayaju I think we can now replace this handling with a "normal" call to `.fetchJson()` here, and just run a "convert to dates" helper function on the success result. It would turn this `Error` into an `HttpError`. Agreed?

File: packages/liveblocks-react/src/umbrella-store.ts [hunk 4]
```diff
@@ -1550,6 +1619,52 @@ export class UmbrellaStore<M extends BaseMetadata> {
...
+    const queryKey = makeVersionsQueryKey(roomId);
+    let resource = this._roomVersions.get(queryKey);
+    if (resource === undefined) {
+      const roomVersionsFetcher = async () => {
+        if (this._client === undefined) {
+          // TODO: Think about other ways to structure this. Throwing a StopRetrying only
+          // makes sense only if we can easily know if the fetcher is going to be wrapped inside the autoRetry function
+          throw new StopRetrying(
+            "Client is required in order to load threads for the room"
+          );
```
Comment (nvie) [inline packages/liveblocks-react/src/umbrella-store.ts:1633]: This should now also be more natural to handle, using the `HttpError` check.

File: packages/liveblocks-react/src/config.ts [hunk 5]
```diff
@@ -11,4 +11,7 @@ export const config = {
 
   USER_THREADS_POLL_INTERVAL: 1 * MINUTES,
   USER_THREADS_MAX_STALE_TIME: 5 * SECONDS,
+
+  HISTORY_VERSIONS_POLL_INTERVAL: 1 * MINUTES,
+  HISTORY_VERSIONS_MAX_STALE_TIME: 3 * SECONDS,
```
Comment (nvie) [inline packages/liveblocks-react/src/config.ts:16]: Any reason to not use 5 seconds here?

File: packages/liveblocks-react/src/room.tsx [hunk 6]
```diff
@@ -2115,23 +2099,42 @@ function useHistoryVersions(): HistoryVersionsAsyncResult {
...
+    //    *next* render after that, a *new* fetch/promise will get created.
+  );
 
   const state = useSyncExternalStoreWithSelector(
     store.subscribe,
     getter,
     getter,
     identity,
-    shallow
+    shallow2
```
Comment (nvie) [inline packages/liveblocks-react/src/room.tsx:2137]: Good catch!

File: packages/liveblocks-react/src/room.tsx [hunk 7]
```diff
@@ -2146,30 +2149,14 @@ function useHistoryVersions(): HistoryVersionsAsyncResult {
...
-    throw getRoomVersions(room);
-  } else if (state.error) {
-    throw state.error;
-  }
-
-  return state;
+  const result = useHistoryVersions();
+  assert(!result.error, "Did not expect error");
+  assert(!result.isLoading, "Did not expect loading");
+  return result;
```
Comment (nvie) [inline packages/liveblocks-react/src/room.tsx:2159]: Love seeing the symmetry here! 😍

File: packages/liveblocks-react/src/umbrella-store.ts [hunk 8]
```diff
@@ -534,7 +595,7 @@ export type UmbrellaStoreState<M extends BaseMetadata> = {
    * Versions by roomId
    * e.g. { 'room-abc': {versions: "all versions"}}
    */
-  versionsByRoomId: Record<string, HistoryVersion[]>;
+  versionsByRoomId: Record<string, Record<string, HistoryVersion>>;
```
Comment (nvie) [inline packages/liveblocks-react/src/umbrella-store.ts:598]: Why did you change this data structure here internally? When we fetch these from the backend, it's a list, and when we expose it from the hook it's also a list. In between we now have to convert the list to a Map, and back to an array again when we read it.

Is this done to remove "duplicate" version IDs? Ideally there would not be such duplicates in the first place, right?

**Review by nvie — APPROVED** [review]
Lovely changes! Great work, @nimeshnayaju! 😍

File: packages/liveblocks-react/src/config.ts (hunk 5 above)
Comment (nimeshnayaju) [reply packages/liveblocks-react/src/config.ts]: This is supposed to be 5 seconds 😅 . It slipped from my example testing. Thanks for pointing out.
